### PR TITLE
chore(shell-api): combine common Cursor + AbstractCursor code

### PR DESCRIPTION
**Pull request by addaleax**
##### chore(shell-api): combine common Cursor + AbstractCursor code

Create an `AbstractCursor` class (named like the driver equivalent)
as a common base class for `Cursor` and `AggregationCursor`
to reduce code duplication.

##### chore(shell-api): allow `@returnType('this')`

This helps with more code deduplication, because it can make wrappers
go away that were only needed to re-specify the return type
in a subclass.

### CODE REVIEW

File: packages/i18n/src/locales/en_US.ts [hunk 1]
```diff
@@ -167,6 +167,12 @@ const translations: Catalog = {
           }
         },
       },
+      AbstractCursor: {
```
Comment (mcasimir) [inline packages/i18n/src/locales/en_US.ts:170]: Why is this necessary here? I don't think we want to make this class visible to the user in any way or?

Can we rather exclude this one with some annotation on the `AbstractCursor` class, something like "@abstract" or "@nohelp"?

Comment (mcasimir) [reply packages/i18n/src/locales/en_US.ts:170]: And of course i've tagged random users ... 🤦

Comment (addaleax) [reply packages/i18n/src/locales/en_US.ts:170]: > Can we rather exclude this one with some annotation on the AbstractCursor class, something like `@abstract` or `@nohelp`?

Yes, let me try that (I think @gribnoysup’s #784 also has the potential to make this kind of thing much easier in the future)

**Review by mcasimir — APPROVED** [review]

File: packages/i18n/src/locales/en_US.ts (hunk 1 above)
Comment (addaleax) [reply packages/i18n/src/locales/en_US.ts]: @mcasimir Okay, done :) Feel free to take another look

**Review by mcasimir — APPROVED** [review]
nice!

File: packages/i18n/src/locales/en_US.ts (hunk 1 above)
Comment (gribnoysup) [reply packages/i18n/src/locales/en_US.ts]: Yeah, totally can take inheritance chains while ignoring those abstract classes into account when generating metadata
